src/dev/flang/ast/Impl.java

```diff
@@ -522,13 +525,42 @@ AbstractType inferredType(Resolution res, AbstractFeature f)
     return switch (_kind)
       {
       case FieldDef    -> _initialValue.typeIfKnown();
-      case RoutineDef  -> _code.typeIfKnown();
+      case RoutineDef  -> typeFromRedefinedFeatures(f);
       case FieldActual -> typeFromInitialValues(res, f, false);
       default -> throw new Error("missing case "+_kind);
       };
   }
 
 
+  AbstractType typeFromRedefinedFeatures(AbstractFeature f)
+  {
+    AbstractType result;
+
+    if (PRECONDITIONS) require
+      (_kind == Kind.RoutineDef);
+
+    var rf = f.redefines()
+              .stream()
+              .filter(x -> x.state().atLeast(State.RESOLVED_TYPES))
+              .map(x -> x.resultType())
+              .filter(x -> !x.isGenericArgument())
+              .filter(x -> x.generics().size() == 0)
+              .filter(x -> !x.isThisType())
+              .collect(Collectors.toList());
+
+    if (rf.size() == 1)
```

Comment:
In case there are several redefined features, your should IMHO still use the type from the redefined feature, at least as long as all are equal (`compareTo()==0`). 
Best would be to have `AbstractType.intersection` (similar to `AbstractType.union`) that returns a type that can be assigned to both (for ref types, the more specific type, not the more general type returned by `union`, and `void ∩ T == void`, `ERROR ∩ T = T`, basically everything the other way around than in `union` without support for boxing (`asRef`)).